packages/resolver/lib/sources/fs.ts

```diff
@@ -23,38 +22,73 @@ export class FS implements ResolverSource {
       }
     }
 
-    try {
-      const result = fs.readFileSync(
-        path.join(searchPath, `${contractName}.json`),
-        "utf8"
-      );
-      return JSON.parse(result);
-    } catch (e) {
-      return null;
-    }
+    const result = this.searchForArtifact(normalizedImportPath, searchPath);
+    return result;
   }
 
-  getContractName(
+  searchForArtifact(
     sourcePath: string,
     searchPath = this.contractsBuildDirectory
   ) {
+    // most cases we can use the `{name}.sol -> {name}.json` convention
+    // console.log(sourcePath);
+    // console.log(searchPath);
+    const likelyFileBasename = path.basename(sourcePath, ".sol");
+    // console.log(likelyFileBasename);
+    // console.log(path.join(searchPath, `${likelyFileBasename}.json`));
+    const likelyArtifactString = fs.readFileSync(
+      path.join(searchPath, `${likelyFileBasename}.json`),
+      "utf8"
+    );
+
+    let likelyArtifact;
+    try {
+      if (likelyArtifactString) {
+        likelyArtifact = JSON.parse(likelyArtifactString);
```

Comment:
It's an edge case, for sure, but we should check to make sure that `likelyArtifact.sourcePath` ends with `sourcePath`

Reply:
I had to revert this change because `sourcePath` is not the same as `likelyArtifact.sourcePath`. The naming was confusing, so I updated it to `importedSourcePath` and added a JSDOC to explain that it's the string passed in via `artifacts.require()`.

Having the check would mismatch the following:

`likelyArtifact.sourcePath = "/tmp/tmp-3266431-Ru5Iww0nfJeN/contracts/Migrations.sol"`
`sourcePath /* now known as importedSourcePath */ = "Migrations"`

Reply:
Where this can go wrong is something like this... let's say you have `A.sol` that defines contract `B`, and `B.sol` that defines contract `A`

The existing code IIRC handles this by verifying the guess. In your proposed change, we lose this verification.

This might not seem like a problem in the abstract example, but this sort of thing can come up when people copy and paste multiple contracts into one source file, and multiple artifacts will get generated.

Now, doing this short-circuit check is fine; all I'm saying is that we should look at `likelyArtifact.sourcePath` to make sure it matches what the user entered, before blindly returning it.

Reply:
> The existing code IIRC handles this by verifying the guess. In your proposed change, we lose this verification.

Is there a test you can point me to that demonstrates the existing code handles this?

---

I tried to give this the benefit of the doubt one last time and do something more robust than my prior attempt. I implemented:

``` typescript
      if (likelyArtifactString) {
        likelyArtifact = JSON.parse(likelyArtifactString);

        if (
          path.basename(likelyArtifact.sourcePath, ".sol") ===
            path.basename(importSourcePath, ".sol")
        ) {
          return likelyArtifact;
        }
      }
```

But this broke several `@truffle/debugger` tests. One example is https://github.com/trufflesuite/truffle/blob/develop/packages/debugger/test/context.js#L83

In my code, this would have `importSourcePath === "OuterContract"`. My short circuit would look for `"OuterContract.json"` in the `contractsBuildDirectory`. It would find it and parse it. However `likelyArtifact.sourcePath === "/tmp/tmp-2027082-yBcgRS8DiwlZ/contracts/OuterLibrary.sol"` so the basename would be comparing `OuterContract` to `OuterLibrary`. So my short circuit would fail.

Then we'd move onto my more lengthy lookup, which would search for every `json` file in the directory **except** `OuterContract.json` because we already tried it (so why try it again?). Since we filter out the correct artifact, we'll never find it.

---

However, I can add a check that looks at `likelyArtifact.contractName` instead of `likelyArtifact.sourcePath`? I tried the below code, but i got some `Unknown exit code: 1` errors in tests

``` typescript
      if (likelyArtifactString) {
        likelyArtifact = JSON.parse(likelyArtifactString);

        if (
          path.basename(likelyArtifact.contractName, ".sol") ===
            path.basename(importSourcePath, ".sol")
        ) {
          return likelyArtifact;
        }
      }
```

I believe this is a more correct check.

---

As an aside, instead of removing the `.sol` extension or looking at the `path.basename` of `importSourcePath`, I believe the FS resolver should error when it is provided a filename/path instead of an explicit contract name. Thoughts? Would this break current users?

---

Outside of this, it'd be very helpful if you could provide a failing test that explains the scenario you're talking about since this PR seems to be lower priority for both of us (and therefore we never are working on it at the same time)